pkg/cluster/cluster_info_scraper.go

```diff
@@ -283,36 +241,85 @@ func (s *ClusterScraper) GetPVCs(namespace string, opts metav1.ListOptions) ([]*
+	existing := 0
+	added := 0
+	for _, pod := range pods {
+		podControllerInfoKey := util.PodControllerInfoKey(pod)
+		if _, exists := s.cache.Get(podControllerInfoKey); exists {
+			// Pod's controller info already exists in the cache
+			existing++
+			continue
+		}
+		ownerInfo, err := util.GetPodParentInfo(pod)
+		if err != nil || util.IsOwnerInfoEmpty(ownerInfo) {
+			// Pod does not have controller
+			glog.V(3).Infof("Skip updating controller for pod %v/%v: pod has no controller.",
+				pod.Namespace, pod.Name)
+			continue
+		}
+		controller, exists := controllers[ownerInfo.Uid]
+		if !exists {
+			// Could be custom controller. We do not bulk process custom controller.
+			glog.V(3).Infof("Skip updating controller %v/%v for pod %v/%v: controller not cached.",
+				ownerInfo.Kind, ownerInfo.Name, pod.Namespace, pod.Name)
+			continue
+		}
+		kind := controller.Kind
+		if kind == commonutil.KindReplicationController || kind == commonutil.KindReplicaSet {
+			// Get grandparent of pod
+			gpOwnerInfo, gpOwnerSet := util.GetOwnerInfo(controller.OwnerReferences)
+			if gpOwnerSet {
+				// Found it
+				s.cache.Set(podControllerInfoKey, gpOwnerInfo, 0)
+				added++
+				continue
+			}
+		}
+		s.cache.Set(podControllerInfoKey, ownerInfo, 0)
+		added++
+	}
+	glog.V(2).Infof("Finished updating pod controller cache."+
+		" Total pod scanned: %d, cached: %d, newly added: %d.", len(pods), existing, added)
+}
+
+// GetPodControllerInfo gets grandParent (parent's parent) information of a pod: kind, name, uid
 // If parent does not have parent, then return parent info.
 // Note: if parent kind is "ReplicaSet", then its parent's parent can be a "Deployment"
 //       or if its a "ReplicationController" its parent could be "DeploymentConfig" (as in openshift).
 // The function also returns the retrieved parent and parents crud interface for use by the callers.
-func (s *ClusterScraper) GetPodGrandparentInfo(pod *api.Pod, ignoreCache bool) (string, string,
-	string, *unstructured.Unstructured, dynamic.ResourceInterface, error) {
+// If ignoreCache is set to false, this function will first try to get pod's controller info from the cache
+// maintained by the ClusterScraper.
+func (s *ClusterScraper) GetPodControllerInfo(
+	pod *api.Pod, useCache bool) (util.OwnerInfo, *unstructured.Unstructured, dynamic.ResourceInterface, error) {
 	podControllerInfoKey := util.PodControllerInfoKey(pod)
-	if !ignoreCache {
+	if useCache {
 		// Get pod controller info from cache if exists
 		controllerInfoCache, exists := s.cache.Get(podControllerInfoKey)
 		if exists {
-			controllerInfo, ok := controllerInfoCache.(kubeControllerInfo)
-			if !ok {
-				return "", "", "", nil, nil, fmt.Errorf("error getting controller info cache data: controllerInfoCache is '%t' not 'kubeControllerInfo'",
-					controllerInfoCache)
+			if ownerInfo, ok := controllerInfoCache.(util.OwnerInfo); ok {
+				return ownerInfo, nil, nil, nil
 			}
```

Comment:
Although this is unlikely, as `podControllerInfoKey` exists in the cache, we could add an else condition to record the error if controllerInfoCache is not `OwnerInfo` type like:
```
else {
    return nil, nil, fmt.Errorf("error getting controller info cache data: controllerInfoCache is '%t' not 'OwnerInfo'", controllerInfoCache);
}
```

Reply:
I added a warning message. But I think we should not return here, but keep trying the non-cached approach instead.

Reply:
Yeah, make sense.